design/baremetal-operator/node-specific-kernel-args.md

```diff
@@ -0,0 +1,166 @@
+3. Implement support for the `preProvisionKernelArgs` in the Ironic provisioner
+
+### Non-Goals
+
+1. To implement any sort validation for the new optional field
+2. Providing a default value for the new spec field
+
+## Proposal
+
+This document proposes a new BareMetalHost spec attribute intended for
+configuring the pre-provisioning image's (usually IPA) kernel arguments.
+
+The proposal involves the introduction of a new spec field thus a modification
+to the BMO API. The API change is intended to be backward compatible as the
+new `preProvisionKernelArgs` field is optional.
+
+### Implementation Details/Notes/Constraints
+
+The `preProvisionKernelArgs` field would be located directly under `spec`
+section. The format of the new field would be a list/array of individual
+strings. The new field would be optional and it would be considered
+either a NOOP or a cleaning operation depending on the previous value of the
+field in case it would be not present or have no value or would have an empty
+array as a value.
+
+The proposed BMH looks like the following:
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ preProvisionKernelArgs:
+   - --timeout
+   - 60000
+   - systemd.journald.forward_to_console=yes
+   - ipa-debug=1
+   ...
+   ...
+   ...
+ bootMACAddress: 52:54:00:b7:b2:6f
+...
+...
+...
+
+```
+
+The example contains a snippet of a common IPA kernel argument list.
+The `preProvisionKernelArgs` list would be received by the `Provisioner`
+interface as is. The handling of the list would be up to the particular
+implementations of the `Provisioner` interface. In case of the `Ironic`
+provisioner the elements from the `preProvisionKernelArgs` would be
+concatenated with spaces into a single string such that the string would start
+with the first element (lowest index) from the list and would end with the last
+(highest index) element. The elements between the first and last element would
+be concatenated in an ascending order based on their index number in the array
+with additional spaces between the elements.
+
+Given the BMH snippet from above, the kernel argument list generated by
+the `Ironic provisioner` and posted to the Ironic Node API  would look like
+`--timeout 60000 systemd.journald.forward_to_console=yes ipa-debug=1 ...`.
+
+### Risks and Mitigations
+
+Providing a faulty list kernel arguments might result in boot issues
+or unexpected behavior during the life cycle of the pre-provisioning image.
+
+The user has to be aware that the kernel arguments supplied here will fully or
+partially overwrite the default arguments depending on how the particular
+`Provisioner` is handling the argument list. In case of Ironic provisioner
+the `preProvisionKernelArgs` will completely replace the default argument
+list used by Ironic.
```

Comment:
See above, I'd prefer to avoid a full override. It will require a user to guess Ironic options that must stay present.

Reply:
I have dropped the full override approach for this proposal.